cpp/src/io/utilities/wrapper_utils.hpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "cudf.h"
+#include "utilities/error_utils.hpp"
+
+#include <cuda_runtime.h>
+#include <rmm/rmm.h>
+
+#include <cstring>
+
+/**
+ * @brief A helper class that wraps a gdf_column and any associated memory.
+ *
+ * This abstraction initializes and manages a gdf_column (fields and memory)
+ * while still allowing direct access. Memory is automatically deallocated
+ * unless ownership is transferred via releasing and assigning the raw pointer.
+ **/
+class gdf_column_wrapper {
+ public:
+  gdf_column_wrapper(gdf_size_type size, gdf_dtype dtype,
+                     gdf_dtype_extra_info dtype_info, const std::string name) {
+    col = (gdf_column *)malloc(gdf_column_sizeof());
+    col->col_name = (char *)malloc(name.length() + 1);
+    std::strcpy(col->col_name, name.c_str());
+    gdf_column_view_augmented(col, nullptr, nullptr, size, dtype, 0, dtype_info);
+  }
+
+  ~gdf_column_wrapper() {
+    if (col) {
+      RMM_FREE(col->data, 0);
+      RMM_FREE(col->valid, 0);
+      free(col->col_name);
+    }
+    free(col);
+  };
+
+  gdf_column_wrapper(const gdf_column_wrapper &other) = delete;
+  gdf_column_wrapper(gdf_column_wrapper &&other) : col(other.col) {
+    other.col = nullptr;
+  }
+
+  gdf_error allocate() {
+    // For strings, just store the ptr + length. Eventually, column's data ptr
+    // is replaced with an NvString instance created from these pairs.
+    const auto num_rows = std::max(col->size, 1);
+    const auto column_byte_width = (col->dtype == GDF_STRING)
+                                       ? sizeof(std::pair<const char *, size_t>)
+                                       : gdf_dtype_size(col->dtype);
+
+    RMM_TRY(RMM_ALLOC(&col->data, num_rows * column_byte_width, 0));
+    RMM_TRY(RMM_ALLOC(&col->valid, gdf_valid_allocation_size(num_rows), 0));
+    CUDA_TRY(cudaMemset(col->valid, 0, gdf_valid_allocation_size(num_rows)));
+
+    return GDF_SUCCESS;
+  }
+
+  gdf_column *operator->() const { return col; }
+  gdf_column *get() const { return col; }
+  gdf_column *release() {
+    auto temp = col;
+    col = nullptr;
+    return temp;
+  }
+
+ private:
+  gdf_column *col = nullptr;
+};
+
+/**
+ * @brief A helper class that wraps fixed-length device memory for the GPU, and
```

Comment:
Just to confirm - there is no automatic synchronization between host and device buffers?

Reply:
Correct. For the first iteration, I've just kept the wrapper very simple and had the user perform any explicit synchronization required. This made it easier to port over the original Parquet reader code during development.
We can certainly add implicit sync, or helper methods to sync though.